src/overlaybd/zfile/compressor.cpp

```diff
@@ -192,69 +197,173 @@ class LZ4Compressor : public BaseCompressor {
         return (qat_enable ? DEFAULT_N_BATCH : 1);
     }
 
-    virtual int do_compress(size_t *src_chunk_len, size_t *dst_chunk_len,
-                            size_t dst_buffer_capacity, size_t nblock) override {
+    int do_compress(size_t *src_chunk_len, size_t *dst_chunk_len, size_t dst_buffer_capacity,
+                    size_t nblock) override;
 
-        int ret = 0;
+    int do_decompress(size_t *src_chunk_len, size_t *dst_chunk_len, size_t dst_buffer_capacity,
+                      size_t n) override;
 #ifdef ENABLE_QAT
-        if (qat_enable) {
-            ret = LZ4_compress_qat(pQat, &raw_data[0], src_chunk_len, &compressed_data[0],
-                                   dst_chunk_len, n);
-            if (ret < 0) {
-                LOG_ERROR_RETURN(EFAULT, -1, "LZ4 compress data failed. (retcode: `).", ret);
-            }
-            return ret;
+private:
+    session_ptr get_session();
+    friend struct cached_session_t;
+    std::vector<session_ptr> sessions;
+    std::mutex mutex;
+#endif
+};
+
+#ifdef ENABLE_QAT
+static bool setup_session(LZ4Compressor::session_ptr &session) {
+    int rc;
+    rc = qzInit(session.get(), QZ_SW_BACKUP_DEFAULT);
+    if (rc != QZ_OK && rc != QZ_DUPLICATE)
+        return false;
+    QzSessionParamsLZ4_T params;
+    rc = qzGetDefaultsLZ4(&params);
+    if (rc != QZ_OK)
+        return false;
+    // params.data_fmt = QZ_LZ4;
+    params.common_params.comp_algorithm = QZ_LZ4;
+    params.common_params.comp_lvl = 1;
+    params.common_params.direction = QZ_DIR_BOTH;
+    rc = qzSetupSessionLZ4(session.get(), &params);
+    if (rc != QZ_OK)
+        return false;
+
+    return true;
+}
+
+// put the session back to the session pool in a RAII manner
+struct cached_session_t {
```

Comment:
Use photon::object_cache<T> instead. It is more optimized for concurrency & performance. And it automatically releases objects as they get cold.